cirq/sim/density_matrix_simulator.py

```diff
@@ -168,8 +168,11 @@ def _run(
         resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
         check_all_resolved(resolved_circuit)
 
-        if circuit.are_all_measurements_terminal() and not any(
-            circuit.findall_operations(lambda op: isinstance(op, circuits.CircuitOperation))
+        unmeasured_prefix, general_suffix = split_into_matching_protocol_then_general(
+            resolved_circuit, lambda op: not protocols.is_measurement(op)
+        )
+        if len(general_suffix) <= 1 and not any(
+            general_suffix.findall_operations(lambda op: isinstance(op, circuits.CircuitOperation))
```

Comment:
Almost missed this: applying the previous `circuit`-based check to `general_suffix` should work here. `general_circuit` can be longer than one moment as long as all measurements are terminal and there are no `CircuitOperation`s.

Reply:
fixed